core/src/test/scala/akka/persistence/jdbc/query/EventsByTagMigrationTest.scala

```diff
@@ -0,0 +1,200 @@
+
+  /**
+   * add new column to event_tag table.
+   */
+  def addNewColumn(): Unit = {}
+
+  /**
+   * drop legacy not compatible inserted rows.
+   */
+  def deleteLegacyRows(): Unit = {
+    withStatement { stmt =>
+      stmt.execute(s"""
+                      |DELETE FROM ${tagTableCfg.tableName}
+                      |WHERE ${tagTableCfg.columnNames.persistenceId} IS NULL
+                      |AND ${tagTableCfg.columnNames.sequenceNumber} IS NULL
+                      |""".stripMargin)
+    }
+  }
+
+  /**
+   * drop old FK constraint
+   */
+  def dropLegacyFKConstraint(): Unit = {
+    withStatement(stmt => stmt.execute(s"""ALTER TABLE ${tagTableCfg.tableName} DROP CONSTRAINT "fk_event_journal""""))
+
+  }
+
+  /**
+   * drop old PK  constraint
+   */
+  def dropLegacyPKConstraint(): Unit = {
+    withStatement(stmt => stmt.execute(s"ALTER TABLE ${tagTableCfg.tableName} DROP PRIMARY KEY"))
+  }
+
+  /**
+   * create new PK constraint for PK column.
+   */
+  def addNewPKConstraint(): Unit = {
+    withStatement { stmt =>
+      stmt.execute(s"""
+                      |ALTER TABLE ${tagTableCfg.tableName}
+                      |ADD CONSTRAINT "pk_event_tag"
+                      |PRIMARY KEY (${tagTableCfg.columnNames.persistenceId}, ${tagTableCfg.columnNames.sequenceNumber}, ${tagTableCfg.columnNames.tag})
+                      """.stripMargin)
+    }
+  }
+
+  /**
+   * create new FK constraint for PK column.
+   */
+  def addNewFKConstraint(): Unit = {
+    withStatement { stmt =>
+      stmt.execute(s"""
+                      |ALTER TABLE ${tagTableCfg.tableName}
+                      |ADD CONSTRAINT "fk_event_journal_on_pk"
+                      |FOREIGN KEY (${tagTableCfg.columnNames.persistenceId}, ${tagTableCfg.columnNames.sequenceNumber})
+                      |REFERENCES ${journalTableCfg.tableName} (${journalTableCfg.columnNames.persistenceId}, ${journalTableCfg.columnNames.sequenceNumber})
+                      |ON DELETE CASCADE
+                      """.stripMargin)
+    }
+  }
+
+  /**
+   * alter the event_id to nullable, so we can skip the InsertAndReturn.
+   */
+  def alterEventIdToNullable(): Unit = {
+    withStatement { stmt =>
+      stmt.execute(s"ALTER TABLE ${tagTableCfg.tableName} ALTER COLUMN ${tagTableCfg.columnNames.eventId} BIGINT NULL")
+    }
+  }
+
+  // override this, so we can reset the value.
+  def withRollingUpdateActorSystem(f: ActorSystem => Unit): Unit = {
+    val redundantConfig = redundantConfigOverride.foldLeft(ConfigFactory.load(config)) {
+      case (conf, (path, configValue)) =>
+        conf.withValue(path, configValue)
+    }
+
+    implicit val system: ActorSystem = ActorSystem("migrator-test", redundantConfig)
+    f(system)
+    system.terminate().futureValue
+  }
+
+  it should "migrate event tag to new way" in {
+    // 1. Mock legacy data on here, but actually using redundant write and read.
+    withRollingUpdateActorSystem { implicit system =>
```

Comment:
During the verify step I commented these line here, and using the original approach to insert the same rows (at this time, the table didn't have any new column of journal table PK). By doing this, the outcome aligns more closely with real-world scenarios.

However, in order to avoid redundancy with the old schema, I'm use the new approach here to simulate the insertion.

I've validated this with databases other than H2, as I believe that when using H2, it implies that migration is not necessary.